rfc/src/rfcs/0004-loop-contract-synthesis.md

```diff
@@ -0,0 +1,145 @@
+
+We encourage users to specify such limit while we hardcode some default limit at the same time. 
+
+### Output of Kani when Synthesizer Enable
+When the flag `--synthesize-loop-contracts` is provided, Kani will report different result for different cases
+1. When there exists some loop invariant in the candidate space with which all assertions can be proved, Kani will synthesize the loop contracts, verify the program with the synthesized loop contracts, and report verification SUCCESS;
+2. When there exists no loop invariants in the candidate space that can be used to prove all assertions within the resource limit, Kani will report the verification result with the best-effort-synthesized loop contracts.
+
+A question about how do we print the synthesized loop contracts when users request is discussed in **Open question**.
+
+## Detailed Design
+The synthesizer ```goto-synthesizer``` is implemented in the repository of `CBMC`, takes as input a goto binary, and outputs a new goto binary with the synthesized loop contracts applied. Currently, Kani invoke `goto-instrument` to instrument the goto binary `main.goto` into a new goto binary `main_instrumented.goto`, and then invoke ```cbmc``` on `main_instrumented.goto` to get the verification result. The synthesis will happen between calling `goto-instrument` and calling ```cbmc```. That is, we invoke ```goto-synthesizer``` on ```main_instrumented.goto``` to produce a new goto binary ```main_synthesized.goto```, and then call ```cbmc``` on `main_synthesized.goto` instead. 
+
+When invoking ```goto-synthesizer```, we pass the following parameters to it with the flags built in `goto-synthesizer`:
+- the resource limit of the synthesis;
+- the unwinding numbers of loops.
+Note that the way we identify loops in goto and the way we identify loop in rust are different. So we also need to parse the label-based loop id to the index-based loop id used by goto.
+
+
+The enumerator used in the synthesizer enumerate candidates from the language of the following grammar template
+```
+NT_Bool -> NT_Bool && NT_Bool | NT_int == NT_int 
+            | NT_int <= NT_int | NT_int < NT_int 
+            | SAME_OBJECT(terminals_ptr, terminals_ptr)
+            
+NT_int  -> NT_int + NT_int | terminals_int | LOOP_ENTRY(terminals_int)
+            | POINTER_OFFSET(terminals_ptr) | OBJECT_SIZE(terminals_ptr)
+            | POINTER_OFFSET(LOOP_ENTRY(terminals_ptr)) | 1
+```
+where `terminals_ptr` are all pointer variables in the scope, and `terminal_int` are all integers variables in the scope.
+
+## Rationale and alternatives
+
+- Different way to identify loops. Instead of using labels for loops, we can use loop number as what CBMC does. Loop numbers are indexes of loops according to the syntactic order. Also, we can alternatively use inlined pragmas to specify unwinding numbers for loops . However, inlined pragmas will hurt the user experience of no annotations in users code while using Kani.
```

Comment:
Using labels for loops might be a major concern. Can a Rust developer add labels to all loops in their code? What if the loop comes from a dependency?